File: Java/050.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Problem050 {

    public static List<Integer> sieveOfEratosthenes(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = false;
        isPrime[1] = false;

        for (int p = 2; p * p <= limit; p++) {
            if (isPrime[p]) {
                for (int i = p * p; i <= limit; i += p) {
                    isPrime[i] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= limit; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }

    public static int findConsecutive(List<Integer> primes, int bound) {
        Set<Integer> primeSet = new HashSet<>(primes);
        long[] runningSum = new long[primes.size() + 1];
        runningSum[0] = 0;
        for (int i = 0; i < primes.size(); i++) {
            runningSum[i + 1] = runningSum[i] + primes.get(i);
        }

        int maxLen = 0;
        int resultPrime = 0;

        for (int i = 0; i < runningSum.length; i++) {
            for (int j = i + 1; j < runningSum.length; j++) {
                long sum = runningSum[j] - runningSum[i];
                int len = j - i;

                if (sum > bound) {
                    break; // Sum exceeds bound, no need to check further for this i
                }

                if (len > maxLen && primeSet.contains((int) sum)) {
                    maxLen = len;
                    resultPrime = (int) sum;
                }
            }
        }
        return resultPrime;
    }

    public static void main(String[] args) {
        int bound = 1_000_000;
        List<Integer> primes = sieveOfEratosthenes(bound);
        int primeSum = findConsecutive(primes, bound);
        System.out.println("Prime which is the sum of the most consecutive primes: " + primeSum);
    }
}

File: Rust/050.rs
use std::collections::HashSet;

fn sieve_of_eratosthenes(limit: usize) -> Vec<u32> {
    let mut is_prime = vec![true; limit + 1];
    if limit >= 0 { is_prime[0] = false; }
    if limit >= 1 { is_prime[1] = false; }

    for p in 2..=limit {
        if p * p > limit { // Optimization: only iterate up to sqrt(limit)
            break;
        }
        if is_prime[p] {
            for multiple in (p * p..=limit).step_by(p) {
                is_prime[multiple] = false;
            }
        }
    }

    let mut primes = Vec::new();
    for (i, &prime) in is_prime.iter().enumerate() {
        if prime {
            primes.push(i as u32);
        }
    }
    primes
}

fn find_consecutive(primes: &[u32], bound: u32) -> u32 {
    let prime_set: HashSet<u32> = primes.iter().cloned().collect();
    let mut running_sum: Vec<u64> = vec![0; primes.len() + 1];
    for i in 0..primes.len() {
        running_sum[i + 1] = running_sum[i] + primes[i] as u64;
    }

    let mut max_len = 0;
    let mut result_prime = 0;

    for i in 0..running_sum.len() {
        for j in (i + 1)..running_sum.len() {
            let sum = running_sum[j] - running_sum[i];
            let len = j - i;

            if sum > bound as u64 {
                break;
            }

            if len > max_len && prime_set.contains(&(sum as u32)) {
                max_len = len;
                result_prime = sum as u32;
            }
        }
    }
    result_prime
}

fn main() {
    let bound = 1_000_000;
    let primes = sieve_of_eratosthenes(bound);
    let prime_sum = find_consecutive(&primes, bound);
    println!("Prime which is the sum of the most consecutive primes: {}", prime_sum);
}
